content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,110 @@
+
+If you fix the seed at the beginning of your main script for reproducibility and then pass the same RNG to each process to be run in parallel, most of the time this will not give you what you want as this RNG will be deep copied. The same results will thus be produced by each process. One of the solutions is to create as many RNGs as parallel processes with a different seed for each of these RNGs. The issue now is that you cannot choose the seeds as easily as you would think. When you choose two different seeds to instantiate two different RNGs how do you know that the numbers produced by these RNGs will appear as statistically independent? The design of independent RNGs for parallel processes has been an important research question. You can for instance refer to the paper [Random numbers for parallel computers: Requirements and methods, with emphasis on GPUs](https://www.sciencedirect.com/science/article/pii/S0378475416300829) by L'Ecuyer et al. (2017) for a good summary of the different methods on this topic.
+
+From numpy 1.17, it is now very easy to instantiate independent RNGs. Depending on the RNG that you use, different strategies are easily available as documented in the [Parallel random number generation section](https://numpy.org/doc/1.18/reference/random/parallel.html) of the numpy documentation. One of the strategies is to use `SeedSequence` which is an algorithm that makes sure that a not so good user-provided seed results in a good initial state for the RNG. Additionally, it ensures that two close seeds will result in two very different initial states for the RNG that are, with very high probability, independent of each other. You can refer to the documentation of [SeedSequence Spawning](https://numpy.org/doc/1.18/reference/random/parallel.html#seedsequence-spawning) for an example on how to generate independent RNGs from a user-provided seed. I here show how you can do this from an already existing RNG and apply it to the [joblib example](https://joblib.readthedocs.io/en/latest/auto_examples/parallel_random_state.html#fixing-the-random-state-to-obtain-deterministic-results) mentioned above.
+
+
+```python
+import numpy as np
+from joblib import Parallel, delayed
+
+def stochastic_function(seed, high=10):
+    rng = np.random.default_rng(seed)
+    return rng.integers(high, size=5)
+
+seed = 98765
+# create the RNG that you want to pass around
+rng = np.random.default_rng(seed)
+# get the SeedSequence of the passed RNG
+ss = rng.bit_generator._seed_seq
```

Comment:
I would add some very clear+visible warning that this is using a private API. In general we really try not to show usage of private methods as there is no guarantee that we won't change these (and we want to be free to do so).

Reply:
Why not start from a `SeedSequence`?

Reply:
for this example I agree that we could start from a `SeedSequence` but when this occurs at any place in a larger python code, and you are only given the passed RNG you need to get the seed sequence of that RNG.

Reply:
I am not sure why the use of private API here.  I think this was public for a long time (although maybe necessary for the legacy mtrand?)

We also implemnted `rng.spawn(1)` and `bitgen.spawn(1)` more recently, although I think allowing `child = rng.spawn()` (no number to get a single one) might be a nice extension.

Reply:
thanks @seberg for the suggestion! I was not aware of the spawn method, just saw it's in numpy >=1.25, good to know!

I checked for the _seed_seq attribute quickly in the doc but could not find anything (https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.bit_generator.html#numpy.random.Generator.bit_generator), am I missing something?

While you are I actually have another question regarding seeding: it's written in the doc of the random module that it is good to have a large number as a seed to have a unique number different than anyone else. But it is also said in the bit generator doc that a seed lower than 32 bits means that there are some initial states that are impossible to reach. Is that the case even with seed sequence? I thought seed sequence was supposed to create good initial states from bad seed but as I saw on the PCG website I guess seed sequence cannot do everything :).

@stefanv I can open another PR to improve this if that's possible?

Reply:
@albertcthomas You are welcome to modify your blog post!

Reply:
> ed lower than 32 bits means that there are some initial states that are impossible to reach. Is that the case even with seed sequence? I thought seed sequence was supposed to create good initial states from bad seed but as I saw on the PCG website I guess seed sequence cannot do everything :).

I am not the person who knows such details without research, Robert Kern would be.  Seedsequence mixes well, maybe.  But if the initial states has more than 32bit of space, then you need more than 32bits of state to reach it all, maybe?

FWIW, I sometimes use:
```
secrets.randbits(128)
```
to just find a perfect seed in the docs (can throw in a `hex` if you like).